File: Sources/XMTP/Codecs/GroupChatTitleChangedCodec.swift
//
//  GroupChatTitleChangedCodec.swift
//
//
//  Created by Pat Nakajima on 6/11/23.
//

import Foundation

public let ContentTypeGroupTitleChangedAdded = ContentTypeID(authorityID: "xmtp.org", typeID: "groupChatTitleChanged", versionMajor: 1, versionMinor: 0)

public struct GroupChatTitleChanged: Codable {
	// The new title
	public var newTitle: String
}

public struct GroupChatTitleChangedCodec: ContentCodec {
	public var contentType = ContentTypeGroupTitleChangedAdded

	public func encode(content: GroupChatTitleChanged) throws -> EncodedContent {
		var encodedContent = EncodedContent()

		encodedContent.type = ContentTypeGroupTitleChangedAdded
		encodedContent.content = try JSONEncoder().encode(content)

		return encodedContent
	}

	public func decode(content: EncodedContent) throws -> GroupChatTitleChanged {
		let titleChanged = try JSONDecoder().decode(GroupChatTitleChanged.self, from: content.content)
		return titleChanged
	}
}

File: library/src/main/java/org/xmtp/android/library/codecs/GroupChatTitleChangedCodec.kt
package org.xmtp.android.library.codecs

import com.google.gson.GsonBuilder
import com.google.protobuf.kotlin.toByteStringUtf8

val ContentTypeGroupTitleChangedAdded = ContentTypeIdBuilder.builderFromAuthorityId(
    "xmtp.org",
    "groupChatTitleChanged",
    versionMajor = 1,
    versionMinor = 0
)

// The new title
data class GroupChatTitleChanged(var newTitle: String)

data class GroupChatTitleChangedCodec(override var contentType: ContentTypeId = ContentTypeGroupTitleChangedAdded) :
    ContentCodec<GroupChatTitleChanged> {

    override fun encode(content: GroupChatTitleChanged): EncodedContent {
        val gson = GsonBuilder().create()
        return EncodedContent.newBuilder().also {
            it.type = ContentTypeGroupTitleChangedAdded
            it.content = gson.toJson(content).toByteStringUtf8()
        }.build()
    }

    override fun decode(content: EncodedContent): GroupChatTitleChanged {
        val gson = GsonBuilder().create()
        return gson.fromJson(content.content.toStringUtf8(), GroupChatTitleChanged::class.java)
    }
}
